File: npmjs/pray-cli/src/cli/suggest.ts
export const TOP_LEVEL_COMMANDS = [
  "add",
  "apply",
  "clean",
  "confess",
  "drift",
  "explain",
  "format",
  "help",
  "init",
  "install",
  "list",
  "login",
  "manifest",
  "outdated",
  "package",
  "plan",
  "prayer",
  "publish",
  "remove",
  "render",
  "repo",
  "serve",
  "sync",
  "tree",
  "trust",
  "unlock",
  "update",
  "vendor",
  "verify",
  "version",
] as const;

export function unknownCommandMessage(command: string): string {
  let message = `unknown command: ${command}`;
  const suggestion = suggestCommand(command, TOP_LEVEL_COMMANDS);
  if (suggestion) {
    message += `\nDid you mean \`${suggestion}\`?`;
  }
  return message;
}

export function suggestCommand(
  input: string,
  candidates: readonly string[],
): string | undefined {
  const maximumDistance = input.length <= 3 ? 1 : 2;
  let best: { candidate: string; distance: number } | undefined;
  for (const candidate of candidates) {
    const distance = levenshteinDistance(input, candidate);
    if (distance > maximumDistance) {
      continue;
    }
    if (!best || distance < best.distance) {
      best = { candidate, distance };
    }
  }
  return best?.candidate;
}

function levenshteinDistance(left: string, right: string): number {
  const leftChars = [...left];
  const rightChars = [...right];
  const leftLength = leftChars.length;
  const rightLength = rightChars.length;
  if (leftLength === 0) {
    return rightLength;
  }
  if (rightLength === 0) {
    return leftLength;
  }

  let previousRow = Array.from({ length: rightLength + 1 }, (_, index) => index);
  let currentRow = Array.from({ length: rightLength + 1 }, () => 0);

  for (let leftIndex = 0; leftIndex < leftLength; leftIndex += 1) {
    currentRow[0] = leftIndex + 1;
    for (let rightIndex = 0; rightIndex < rightLength; rightIndex += 1) {
      const leftCharacter = leftChars[leftIndex];
      const rightCharacter = rightChars[rightIndex];
      if (leftCharacter === undefined || rightCharacter === undefined) {
        continue;
      }
      const substitutionCost = leftCharacter === rightCharacter ? 0 : 1;
      const deleteCost = (previousRow[rightIndex + 1] ?? 0) + 1;
      const insertCost = (currentRow[rightIndex] ?? 0) + 1;
      const substituteCost = (previousRow[rightIndex] ?? 0) + substitutionCost;
      currentRow[rightIndex + 1] = Math.min(deleteCost, insertCost, substituteCost);
    }
    [previousRow, currentRow] = [currentRow, previousRow];
  }

  return previousRow[rightLength] ?? 0;
}

File: rubygems/pray-cli/lib/pray/cli/suggest.rb
# frozen_string_literal: true

module Pray
  module CLI
    module Suggest
      TOP_LEVEL_COMMANDS = %w[
        add apply clean confess drift explain format help init install list login manifest
        outdated package plan prayer publish remove render repo serve sync tree trust unlock
        update vendor verify version
      ].freeze

      module_function

      def unknown_command_message(command)
        message = "unknown command: #{command}"
        suggestion = suggest_command(command, TOP_LEVEL_COMMANDS)
        suggestion ? "#{message}\nDid you mean `#{suggestion}`?" : message
      end

      def suggest_command(input, candidates)
        maximum_distance = input.length <= 3 ? 1 : 2
        candidates
          .map { |candidate| [candidate, levenshtein_distance(input, candidate)] }
          .select { |(_, distance)| distance <= maximum_distance }
          .min_by { |(_, distance)| distance }
          &.first
      end

      def levenshtein_distance(left, right)
        left_chars = left.chars
        right_chars = right.chars
        left_length = left_chars.length
        right_length = right_chars.length
        return right_length if left_length.zero?
        return left_length if right_length.zero?

        previous_row = (0..right_length).to_a
        current_row = Array.new(right_length + 1, 0)

        left_chars.each_with_index do |left_character, left_index|
          current_row[0] = left_index + 1
          right_chars.each_with_index do |right_character, right_index|
            substitution_cost = left_character == right_character ? 0 : 1
            current_row[right_index + 1] = [
              previous_row[right_index + 1] + 1,
              current_row[right_index] + 1,
              previous_row[right_index] + substitution_cost
            ].min
          end
          previous_row, current_row = current_row, previous_row
        end

        previous_row[right_length]
      end
    end
  end
end
